distributed/batched.py

```diff
@@ -38,87 +39,76 @@ class BatchedSend(object):
     def __init__(self, interval, loop=None):
         self.loop = loop or IOLoop.current()
         self.interval = interval / 1000.
-        self.last_transmission = 0
+
+        self.waker = locks.Event()
+        self.stopped = locks.Event()
+        self.please_stop = False
         self.buffer = []
         self.stream = None
-        self.last_payload = []
-        self.last_send = gen.sleep(0)
         self.message_count = 0
         self.batch_count = 0
+        self.next_deadline = None
 
     def start(self, stream):
         self.stream = stream
-        if self.buffer:
-            self.send_next()
+        self.loop.add_callback(self._background_send)
 
     def __str__(self):
         return '<BatchedSend: %d in buffer>' % len(self.buffer)
 
     __repr__ = __str__
 
     @gen.coroutine
-    def send_next(self, wait=True):
-        try:
-            now = default_timer()
-            if wait:
-                wait_time = min(self.last_transmission + self.interval - now,
-                                self.interval)
-                yield gen.sleep(wait_time)
-            yield self.last_send
-            self.buffer, payload = [], self.buffer
-            self.last_payload = payload
-            self.last_transmission = now
+    def _background_send(self):
+        while not self.please_stop:
+            with ignoring(gen.TimeoutError):
+                yield self.waker.wait(self.next_deadline)
+                self.waker.clear()
+            if not self.buffer:
+                # Nothing to send
+                self.next_deadline = None
+                continue
+            if (self.next_deadline is not None and
+                self.loop.time() < self.next_deadline):
+                # Send interval not expired yet
+                continue
+            payload, self.buffer = self.buffer, []
             self.batch_count += 1
-            self.last_send = write(self.stream, payload)
-        except Exception as e:
-            logger.exception(e)
-            raise
+            try:
+                yield write(self.stream, payload)
+            except Exception:
+                logger.exception("Error in batched write")
+                break
+            self.next_deadline = self.loop.time() + self.interval
```

Comment:
Then the `yield write` wasn't really useful in the previous version? There's no need to wait on the write future if we want to base the deadline on the start of the write operation.

Related question: what is with `gen.with_timeout(timedelta(seconds=0.01),...)` in `d.core.write`?